metals/src/main/scala/scala/meta/internal/metals/callHierarchy/FindCallsResult.scala

```diff
@@ -0,0 +1,152 @@
+    definitionNameRange: lsp4j.Range,
+    fromRanges: List[lsp4j.Range],
+)(implicit ec: ExecutionContext)
+    extends FindCallsResult[
+      lsp4j.CallHierarchyIncomingCall,
+      FindIncomingCallsResult,
+    ] {
+
+  override protected def aggregate(
+      updatedFromRanges: List[lsp4j.Range]
+  ): FindIncomingCallsResult =
+    this.copy(fromRanges = updatedFromRanges)
+
+  private val anonToLSP: (AbsolutePath, TextDocument) => (
+      CallHierarchyItemBuilder,
+      Array[String],
+      CancelToken,
+  ) => Future[lsp4j.CallHierarchyIncomingCall] =
+    toLSP((from, fromRanges) =>
+      new lsp4j.CallHierarchyIncomingCall(from, fromRanges)
+    ) _
+
+  def toLSP(
+      source: AbsolutePath,
+      doc: TextDocument,
+      builder: CallHierarchyItemBuilder,
+      visited: Array[String],
+      token: CancelToken,
+  ): Future[lsp4j.CallHierarchyIncomingCall] =
+    anonToLSP(source, doc)(builder, visited, token)
+}
+
+/**
+ * Result of incoming calls finder.
+ *
+ * @param canBuildHierarchyItem The item that is called wrapped in CanBuildHierarchyItem.
+ * @param definitionNameRange The range that should be selected.
+ * @param fromRanges Ranges at which the item is called.
+ * @param source AbsolutePath where the item that is called is defined.
+ * @param doc TextDocument where the item that is called is defined.
+ */
+private[callHierarchy] case class FindOutgoingCallsResult(
+    occurence: SymbolOccurrence,
+    definitionNameRange: lsp4j.Range,
+    fromRanges: List[lsp4j.Range],
+    source: AbsolutePath,
+    doc: TextDocument,
+)(implicit ec: ExecutionContext)
+    extends FindCallsResult[
+      lsp4j.CallHierarchyOutgoingCall,
+      FindOutgoingCallsResult,
+    ] {
+
+  override protected def aggregate(
+      updatedFromRanges: List[lsp4j.Range]
+  ): FindOutgoingCallsResult =
+    this.copy(fromRanges = updatedFromRanges)
+
+  private val anonToLSP
+      : (CallHierarchyItemBuilder, Array[String], CancelToken) => Future[
+        lsp4j.CallHierarchyOutgoingCall
+      ] =
+    toLSP((to, fromRanges) =>
+      new lsp4j.CallHierarchyOutgoingCall(to, fromRanges)
+    )(source, doc)
+
+  def toLSP(
+      builder: CallHierarchyItemBuilder,
+      visited: Array[String],
+      token: CancelToken,
+  ): Future[lsp4j.CallHierarchyOutgoingCall] =
+    anonToLSP(builder, visited, token)
+}
+
+private[callHierarchy] object FindCallsResult {
+  private def clearRanges(ranges: List[lsp4j.Range]): List[lsp4j.Range] = {
+    val distinctedRanges = ranges.toSet
+    distinctedRanges
+      .filterNot(range => (distinctedRanges - range).exists(_.encloses(range)))
+      .toList
+  }
+
+  /** Aggregate results of incoming calls finder by grouping the fromRanges. */
+  def group[T <: FindCallsResult[_, T]](
+      results: List[T]
+  ): List[T] =
+    results.groupBy(_.occurence).values.toList.collect {
+      case result @ first :: _ =>
+        first.aggregate(clearRanges(result.flatMap(_.fromRanges)))
+    }
```

Comment:
This needs more explanation what's going on. If you answer and explain to me why this is needed in this thread, we can come up with some scaladoc together ;)

Reply:
Basically, when we search for calls, we needs to find the ranges of the calls (the `fromRanges` attribute), and after we need to group the result that have the same CallHierarchyItem.

Example: 

```scala
def foo  = {
  a() //1
  a() //2
}
```
If we compute the outgoings call of `foo` We will have as results , something like : 
```
[
{to: CallHierarchyItem that represent the definition of a(), fromRanges: [range of the first a() call] }, 
{to: CallHierarchyItem that represent the definition of a(), fromRanges: [range of the second a() call] },
]
```
So we need to group them to have:
```
[{to: CallHierarchyItem that represent the definition of a(), fromRanges: [range of the first a() call, range of the second a() call]}]
```

Finally the goal of clearRanges is to avoid the potential duplication of calls caused by the synthetics symbols by just make sure to not have a range that encloses an other.

It that clear ?